### PR TITLE
Add Stax UI code.

### PR DESCRIPTION
**Pull request by agrojean-ledger**

### CODE REVIEW

File: .github/workflows/coding_style_checks.yml [hunk 1]
```diff
@@ -16,4 +16,15 @@ jobs:
...
\ No newline at end of file
+      version: 11
+
+  tests_lint_check:
+    name: Check tests linting
+    runs-on: ubuntu-latest
+    steps:
+    - name: Clone
+      uses: actions/checkout@v3
+    - run: pip install yapf toml
```
Comment (xchapron-ledger) [inline .github/workflows/coding_style_checks.yml]:
```suggestion
    - run: pip install yapf
```

File: .gitignore [hunk 2]
```diff
@@ -1,10 +1,9 @@
 .DS_Store
 
-/bin/
-/debug/
-/dep/
-/obj/
+bin/
+debug/
+dep/
+obj/
```
Comment (xchapron-ledger) [inline .gitignore]:
```suggestion
```

File: Makefile [hunk 3]
```diff
@@ -1,140 +1,88 @@
...
+#     Application custom permissions   #
+########################################
+# See SDK `include/appflags.h` for the purpose of each permission
+HAVE_APPLICATION_FLAG_BOLOS_SETTINGS = 1
+HAVE_APPLICATION_FLAG_GLOBAL_PIN = 1
 
-ifeq ($(GCCPATH),)
-$(info GCCPATH is not set: arm-none-eabi-* will be used from PATH)
-endif
+APPNAME = "NEO"
```
Comment (xchapron-ledger) [inline Makefile]: Maybe could be above near version and all?

File: Makefile [hunk 4]
```diff
@@ -1,140 +1,88 @@
...
-
-# import rules to compile glyphs(/pone)
-include $(BOLOS_SDK)/Makefile.glyphs
-
-# Import generic rules from the SDK
-include $(BOLOS_SDK)/Makefile.rules
-
-listvariants:
-	@echo VARIANTS neo
+include $(BOLOS_SDK)/Makefile.standard_app
```
Comment (xchapron-ledger) [inline Makefile]:
```suggestion
include $(BOLOS_SDK)/Makefile.standard_app

```
Missing empty line at the end of file?

File: glyphs/round_warning_64px.gif
Comment (xchapron-ledger) [inline glyphs/round_warning_64px.gif:1]: You should use the exact same glyph from the SDK

File: glyphs/wallet_64px.gif
Comment (xchapron-ledger) [inline glyphs/wallet_64px.gif:1]: Could be added in the SDK maybe?

File: src/main.c [hunk 5]
```diff
@@ -126,6 +66,8 @@ static void neo_main(void) {
     // sure the io_event is called with a
     // switch event, before the apdu is replied to the bootloader. This avoid
     // APDU injection faults.
+
+    PRINTF("ENTER MAIN \n");
```
Comment (xchapron-ledger) [inline src/main.c]: should we keep this debug?

File: src/main.c [hunk 6]
```diff
@@ -233,33 +172,35 @@ static void neo_main(void) {
...
-                                                   BIP44_PATH_LEN,
-                                                   privateKeyData,
-                                                   NULL);
+                        unsigned char privateKeyData[64];
+
+                        if (os_derive_bip32_no_throw(CX_CURVE_256R1,
+                                                     bip44_path,
+                                                     BIP44_PATH_LEN,
+                                                     privateKeyData,
+                                                     NULL)) {
```
Comment (xchapron-ledger) [inline src/main.c]: Maybe:
```suggestion
                        if (os_derive_bip32_no_throw(CX_CURVE_256R1,
                                                     bip44_path,
                                                     BIP44_PATH_LEN,
                                                     privateKeyData,
                                                     NULL) != CX_OK) {
```

File: src/main.c [hunk 7]
```diff
@@ -233,33 +172,35 @@ static void neo_main(void) {
...
+                            THROW(0x6D00);
+                        }
                         cx_ecdsa_init_private_key(CX_CURVE_256R1, privateKeyData, 32, &privateKey);
 
                         // generate the public key.
                         cx_ecdsa_init_public_key(CX_CURVE_256R1, NULL, 0, &publicKey);
-                        cx_ecfp_generate_pair(CX_CURVE_256R1, &publicKey, &privateKey, 1);
+                        cx_ecfp_generate_pair_no_throw(CX_CURVE_256R1, &publicKey, &privateKey, 1);
```
Comment (xchapron-ledger) [inline src/main.c]: this whole block could be changed to a simple call to `bip32_derive_get_pubkey_256` from `ledger-secure-sdk/lib_standard_app/crypto_helpers.h`

File: src/main.c [hunk 8]
```diff
@@ -306,15 +250,10 @@ static void neo_main(void) {
                         cx_sha256_t pubKeyHash;
                         cx_sha256_init(&pubKeyHash);
 
-                        cx_hash(&pubKeyHash.header, CX_LAST, publicKey.W, 65, result, 32);
-                        tx += cx_ecdsa_sign((void *) &privateKey,
```
Comment (xchapron-ledger) [inline src/main.c:310]: Do we really need to drop the signature here?

File: src/main.c [hunk 9]
```diff
@@ -19,49 +19,16 @@
...
-
-static void Timer_Restart() {
-    if (exit_timer != MAX_EXIT_TIMER) {
-        Timer_Set();
-    }
-}
-
-static bool Timer_Expired() {
-    return exit_timer <= 0;
-}
```
Comment (xchapron-ledger) [inline src/main.c:60]: What was the purpose of this timer feature? Are we loosing something here?

File: src/ui.c [hunk 10]
```diff
@@ -666,17 +686,117 @@ void ui_public_key_2(void) {
...
+        ui_idle();
+    }
+}
+
+static bool infoNavCallback(uint8_t page, nbgl_pageContent_t *content) {
+    if (page == 0) {
+        content->type = INFOS_LIST;
+        content->infosList.nbInfos = NB_INFO_FIELDS;
+        content->infosList.infoTypes = (const char **) infoTypes;
+        content->infosList.infoContents = (const char **) infoContents;
```
Comment (xchapron-ledger) [inline src/ui.c]:
```suggestion
        content->infosList.infoTypes = infoTypes;
        content->infosList.infoContents = infoContents;
```

File: src/ui.c [hunk 11]
```diff
@@ -666,17 +686,117 @@ void ui_public_key_2(void) {
...
+
+static void reviewStart(void) {
+    memset(&fields, 0, sizeof(fields));
+    memset(&infoLongPress, 0, sizeof(infoLongPress));
+
+    infoLongPress.text = "Sign transaction";
+    infoLongPress.longPressText = "Hold to sign";
+    infoLongPress.icon = &C_icon_64px;
+
+    pairList.pairs = (nbgl_layoutTagValue_t *) fields;
```
Comment (xchapron-ledger) [inline src/ui.c]: Why the cast?
```suggestion
    pairList.pairs = fields;
```

File: src/ui.c [hunk 12]
```diff
@@ -666,17 +686,117 @@ void ui_public_key_2(void) {
...
+                            rejectUseCaseChoice);
+}
+#endif
+////////////////////////////////////////////////////////////////////////////////////////////////
 
 /** processes the transaction approval. the UI is only displayed when all of the TX has been sent
  * over for signing. */
-const bagl_element_t *io_seproxyhal_touch_approve(const bagl_element_t *e) {
+const void *io_seproxyhal_touch_approve(const void *e) {
     UNUSED(e);
```
Comment (xchapron-ledger) [inline src/ui.c]: While changing the parameter type, why not simpling dropping it as it is unused?
Function can also rename not to show `io_seproxyhal` prefix.

File: src/ui.c [hunk 13]
```diff
@@ -739,12 +865,14 @@ const bagl_element_t *io_seproxyhal_touch_approve(const bagl_element_t *e) {
...
+#ifdef HAVE_BAGL
     ui_idle();
+#endif
     return 0;  // do not redraw the widget
 }
 
 /** deny signing. */
-static const bagl_element_t *io_seproxyhal_touch_deny(const bagl_element_t *e) {
+static const void *io_seproxyhal_touch_deny(const void *e) {
     UNUSED(e);
```
Comment (xchapron-ledger) [inline src/ui.c]: Same as above.

File: .github/workflows/coding_style_checks.yml (hunk 1 above)
Comment (agrojean-ledger) [reply .github/workflows/coding_style_checks.yml]: Done

File: .gitignore (hunk 2 above)
Comment (agrojean-ledger) [reply .gitignore]: Done.

File: Makefile (hunk 3 above)
Comment (agrojean-ledger) [reply Makefile]: Done.

File: src/main.c (hunk 5 above)
Comment (agrojean-ledger) [reply src/main.c]: Nope. Will remove it.

File: src/main.c (hunk 9 above)
Comment (agrojean-ledger) [reply src/main.c:60]: I think it was for blue. The tests are passing without this... I have no idea what it was used for.

File: Makefile (hunk 4 above)
Comment (agrojean-ledger) [reply Makefile]: Ok

File: src/main.c (hunk 9 above)
Comment (xchapron-ledger) [reply src/main.c:60]: It looks like a way to timeout UI flows. But as we don't have this on other apps, this is probably not really necessary.

File: glyphs/round_warning_64px.gif
Comment (agrojean-ledger) [reply glyphs/round_warning_64px.gif:1]: Done.

File: glyphs/wallet_64px.gif
Comment (agrojean-ledger) [reply glyphs/wallet_64px.gif:1]: If there is an SDK API that would use it yes why not... Otherwise I won't add it.

File: src/main.c (hunk 6 above)
Comment (agrojean-ledger) [reply src/main.c]: Indeed. I should explicitly check a return value.

File: src/main.c (hunk 7 above)
Comment (agrojean-ledger) [reply src/main.c]: Done.

File: src/main.c (hunk 8 above)
Comment (agrojean-ledger) [reply src/main.c:310]: Nope I'll put it back.

File: src/ui.c (hunk 10 above)
Comment (agrojean-ledger) [reply src/ui.c]: Done.

File: src/ui.c (hunk 11 above)
Comment (agrojean-ledger) [reply src/ui.c]: Done.

File: src/ui.c (hunk 13 above)
Comment (agrojean-ledger) [reply src/ui.c]: Done.

File: src/ui.c (hunk 12 above)
Comment (agrojean-ledger) [reply src/ui.c]: Done.

File: Makefile [hunk 14]
```diff
@@ -1,140 +1,91 @@
...
+ENABLE_BLUETOOTH = 1
+
+########################################
+#         NBGL custom features         #
+########################################
+ENABLE_NBGL_QRCODE = 1
+
+# Use only specific files from standard app
+DISABLE_STANDARD_APP_FILES = 1
+APP_SOURCE_FILES += ${BOLOS_SDK}/lib_standard_app/io.
```
Comment (xchapron-ledger) [inline Makefile]:
```suggestion
APP_SOURCE_FILES += ${BOLOS_SDK}/lib_standard_app/io.c
```

File: src/main.c [hunk 15]
```diff
@@ -212,10 +153,7 @@ static void neo_main(void) {
 
                         // we're asked for the public key.
                     case INS_GET_PUBLIC_KEY: {
-                        Timer_Restart();
-
                         cx_ecfp_public_key_t publicKey;
```
Comment (xchapron-ledger) [inline src/main.c]: You could allocate a buffer` uint8_t raw_pubkey[65]` instead of this maybe?

File: src/main.c [hunk 16]
```diff
@@ -279,25 +214,28 @@ static void neo_main(void) {
...
-                                                   NULL);
+                        unsigned char privateKeyData[64];
+                        if (os_derive_bip32_no_throw(CX_CURVE_256R1,
+                                                     bip44_path,
+                                                     BIP44_PATH_LEN,
+                                                     privateKeyData,
+                                                     NULL)) {
+                            THROW(0x6D00);
+                        }
                         cx_ecdsa_init_private_key(CX_CURVE_256R1, privateKeyData, 32, &privateKey);
```
Comment (xchapron-ledger) [inline src/main.c]:
```suggestion
                        if (bip32_derive_init_privkey_256(CX_CURVE_256R1,
                                                     bip44_path,
                                                     BIP44_PATH_LEN,
                                                     &privateKey,
                                                     NULL) != CX_OK) {
                            THROW(0x6D00);
                        }
```

File: Makefile (hunk 14 above)
Comment (agrojean-ledger) [reply Makefile]: Woups. Thanks.

File: src/main.c (hunk 15 above)
Comment (agrojean-ledger) [reply src/main.c]: Done.

File: src/main.c (hunk 16 above)
Comment (agrojean-ledger) [reply src/main.c]: Done.

File: src/main.c [hunk 17]
```diff
@@ -19,49 +19,17 @@
...
+#include "io.h"
 #include "ui.h"
 #include "neo.h"
+#ifdef HAVE_BAGL
 #include "bagl.h"
-
+#endif
 #define MAX_EXIT_TIMER 4098
 
 #define EXIT_TIMER_REFRESH_INTERVAL 512
```
Comment (xchapron-ledger) [inline src/main.c]: Probably not used anymore?
```suggestion
```

File: src/main.c [hunk 18]
```diff
@@ -357,77 +299,11 @@ static void neo_main(void) {
 }
 
 /** display function */
+#ifdef HAVE_BAGL
 void io_seproxyhal_display(const bagl_element_t *element) {
     io_seproxyhal_display_default((bagl_element_t *) element);
 }
```
Comment (xchapron-ledger) [inline src/main.c]: Already in lib_standard_app/io.c, then can be dropped.

File: src/ui.c [hunk 19]
```diff
@@ -690,29 +810,32 @@ const bagl_element_t *io_seproxyhal_touch_approve(const bagl_element_t *e) {
...
+                                          NULL) != CX_OK) {
+            THROW(0x6D00);
+        }
 
         // Hash is finalized, send back the signature
         unsigned char result[32];
 
-        cx_hash(&hash.header, CX_LAST, G_io_apdu_buffer, 0, result, 32);
+        cx_hash_no_throw(&tx_hash.header, CX_LAST, G_io_apdu_buffer, 0, result, 32);
 #if CX_APILEVEL >= 8
```
Comment (xchapron-ledger) [inline src/ui.c]: IMO we could drop the build for CX_APILEVEL < 8....

File: test/requirements.txt [hunk 20]
```diff
@@ -0,0 +1,4 @@
+eth-keys==0.3.1
+pytest
+ragger[speculos]>=1.6.0
+ragger[ledgerwallet]>=1.6.0
```
Comment (xchapron-ledger) [inline test/requirements.txt]: Could be bumped to 1.9.2

**Review by xchapron-ledger — APPROVED** [review]

File: src/ui.c (hunk 19 above)
Comment (agrojean-ledger) [reply src/ui.c]: Ok let's go

File: test/requirements.txt (hunk 20 above)
Comment (agrojean-ledger) [reply test/requirements.txt]: Ok

File: src/main.c (hunk 17 above)
Comment (agrojean-ledger) [reply src/main.c]: Done

File: src/main.c (hunk 18 above)
Comment (agrojean-ledger) [reply src/main.c]: Done.

**Pull request merged**